autoload/ale/c.vim

```diff
@@ -89,3 +89,59 @@ function! ale#c#FindCompileCommands(buffer) abort
 
     return ''
 endfunction
+
+" Given a buffer number, find a compilation database (compile_commands.json)
+" and parse this into an argument string
+function! ale#c#FindCompileArgs(buffer) abort
+    let l:use_args = []
+
+    let l:compile_commands_path = ale#path#FindNearestFile(a:buffer, 'compile_commands.json')
+    if (!empty(l:compile_commands_path))
+        let l:bufname = fnamemodify(bufname(a:buffer), ':p')
+        let l:command = ''
+        let l:file = ''
+
+        for elem in readfile(l:compile_commands_path)
+            let l:split = split(elem, ':')
```

Comment:
I recommend using the built in JSON decoding functions instead. See `:help json_decode`.

Break this up into two functions, and write tests for each.  Write a function that accepts the result of `json_decode` and takes the options out of it, and a second function which calls the first function. Then you can write a test case which tests finding the file and parsing the options out if it, and a series of tests for the function which doesn't do any I/O.